### ISSUE TITLE
routes ending with `/` are not handled

### ISSUE THREAD

**Issue by NilsJacobsen** [brillout/vite-plugin-ssr#949]
### Description

![image](https://github.com/brillout/vite-plugin-ssr/assets/58360188/3a666043-b7ce-4bd2-92a6-30422105c8f2)

Routes that end with `/` are not handled to redirect to the prior route.

### Expected
`domain.com/docs/` -> `domain.com/docs`

Is that something that you want to integrate by default? I'm sure 99% of the cases this is gonna be wanted.

**Comment by brillout** [comment]
Does it also happen in production?

**Comment by NilsJacobsen** [comment]
Jap also in production.

![image](https://github.com/brillout/vite-plugin-ssr/assets/58360188/09df8505-009c-44e1-9f46-9f34cb18dde2)

**Comment by brillout** [comment]
Can you configure your production server to do the redirection?

I agree it's something that should be built into VPS. I updated https://github.com/brillout/vite-plugin-ssr/issues/926.

**Comment by brillout** [comment]
@martijndirksen @martijnpor I was expecting that someone may need to disable this. Why do you want the trailing slash?

**Comment by martijndirksen** [comment]
We migrated a production platform to vite-plugin-ssr which was already using trailing slashes in its routing. We're left with permanent redirects and we cannot easily go back to a situation without them.

Regardless of our own situation, I feel like enforcing URL conventions is not necessary for vite-plugin-ssr. The users can figure out appropriate solutions. As it stands now it might be too restrictive for framework consumers.

**Comment by brillout** [comment]
I believe a simple workaround is to remove the trailing slash in `renderPage()`:

```js
import { renderPage } from 'vite-plugin-ssr'

server.addMiddleware({
  method: 'GET',
  route: '*', // Catch-all
  async handler(request) {
    let { url } = request
    // Removing trailing slash
    if (url.endsWith('/')) url = url.slice(0, -1)
    const pageContextInit = { urlOriginal: url }
    const pageContext = await renderPage(pageContextInit)
    // ...
  }
})
```

Let me know whether that works out.

I'll consider creating a config if more users need this. Your use case seems too niche to justify a new config.

**Comment by martijndirksen** [comment]
With a slight adjustment, your proposal seems to work. We still use trailing slash redirects (which we do ourselves) and then fool vps by removing that trailing slash when passing in the urlOriginal. Note that we have to check for `/` as we can't pass an empty string to the urlOriginal.

This workaround is a bit nicer than me using `patch-package` to strip out the url normalization from your package, but it still feels unintuitive. I need to place a comment in my source code to explain why we strip the trailing slash again, to avoid developers from removing this and getting redirect loops.

```ts
const pageContextInit: PageContextInit = {
  urlOriginal: req.originalUrl !== '/' && req.originalUrl.endsWith('/') ? req.originalUrl.slice(0, -1) : req.originalUrl,
  // ...
};
```

Our general use case is that our clients have different preferences when it comes to the usage of trailing slashes. Some believe there is a difference in SEO, but there does not seem to be a clear consensus around this. We therefore need to be able to support both with and without. Automatic URL normalization by vite-plugin-ssr would make things more difficult for us in this regard. If you want that kind of stuff from a framework, wouldn't something like Nuxt be a better fit?

**Comment by brillout** [comment]
A new config isn't worth it, if it's used only by a handful of users while there is valid workaround. Duct taping is obviously not ideal, but too many configs isn't either. That's why I'm inclined to create a new config only if I can see a use case for it.

> We therefore need to be able to support both with and without

What's the benefit of disabling trailing slash redirection?

> We still use trailing slash redirects (which we do ourselves)

Why is vite-plugin-ssr's trailing slash redirection causing a problem then?

**Comment by brillout** [comment]
@schaschko If you don't mind elaborating why you need trailing slashes, I'll further consider adding a config for how vite-plugin-ssr should treat trailing slashes.

I can't think of a website I use that has trailing slashes, so I'm inclined to think that your use case is very uncommon.

**Comment by schaschko** [comment]
@brillout Yes of course, please check regarding [trailing slashes](https://developers.google.com/search/blog/2010/04/to-slash-or-not-to-slash#your-sites-configuration-and-your-options). It's a decision whether to serve with or without a trailing slash. Imo it should not be enforced by the VPS.

And I don't think the use case is very uncommon ;)

**Comment by brillout** [comment]
Quickly glazed over your link, but I still don't see the benefit of trailing slashes?

To be clear, I'm very much open for a new config `trailingSlash: true | false | null` (`null` disables the auto redirect).

> And I don't think the use case is very uncommon ;)

In that case someone will quickly be able to explain me why trailing slashes can be usefull and I'll implement the config then ;)

**Comment by schaschko** [comment]
I know what you mean bro, but vps should not bypass the dev imo.

`https://example.com/notthesame` could serve completely different content than `https://example.com/notthesame/` (it won't). As different as `https://example.com/path/to/some/thing` vs. `https://example.com/path/to/some/other/thing`, but we wouldn't automatically "normalize" `/path/to/some/thing` to resolve to `/path/to/some/other/thing`. I know, not the same thing semantically.

In the end, I think the dev should be in control on how to handle urls pointing to different resources, useful or not.

**Comment by MichaelRoosz** [comment]
yes trailing slashes are mostly a cosmetic thing.
but if you are already are using them, then why remove them and add redirects to all your pages? this will impact SEO (redirecting all pages). also some people just prefer to have them.

in my case, having trailing slashes is a hard requirement, not decided by myself, so having such an option would be very helpful. for now I am also messing with "urlOriginal"

**Comment by brillout** [comment]
https://vite-plugin-ssr.com/url-normalization released in `0.4.140`.

**Comment by brillout** [comment]
Thanks for the conversation.

We're looking for sponsors, in particular company sponsorships: https://github.com/sponsors/brillout.
